Add Salesforce Rest Connector example (#1185)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/salesforce-rest-connector/sf-rest-connector-example.md

# Salesforce Rest API Connector Example

The Salesforce REST Connector allows you to work with records in Salesforce, a web-based service that allows organizations to manage contact relationship management (CRM) data. You can use the Salesforce connector to create, query, retrieve, update, and delete records in your organization's Salesforce data. The connector uses the [Salesforce REST API](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_what_is_rest_api.htm) to interact with Salesforce.

## What you'll build

This example explains how to use the Salesforce client to connect with the Salesforce instance and perform the 
following operations:

* Create an Account 
* Execute a SOQL query to retrieve the Account Name and ID in all the existing Accounts. 

> <img src="/assets/img/connectors/Salesforce.png" title="Using Salesforce Rest Connector" width="800" alt="Using Salesforce Rest Connector"/>

Reviewer: Better add some explanation of this diagram as well.

Author: Done

## Configure the connector in WSO2 Integration Studio

> 1. Follow these steps to set up the ESB Solution Project and the Connector Exporter Project. 

Reviewer: Update this so that it is similar to the File Connector example. Made some changes to that so use that as a template.

Author: Done

{!references/connectors/importing-connector-to-integration-studio.md!}

2. Right click on the created ESB Solution Project and select, -> **New** -> **Rest API** to create the REST API. 
    <img src="/assets/img/connectors/adding-an-api.png" title="Adding a Rest API" width="800" alt="Adding a Rest API"/>

> 3. Follow these steps to [generate the Access Tokens for Salesforce](../salesforce-connector/sf-access-token-generation.md) and obtain the Client Id, Client Secret, Access Token and Refresh Token. 

Reviewer: ```suggestion
3. Follow these steps to [generate the Access Tokens for Salesforce](../salesforce-connector/sf-access-token-generation.md) and obtain the Client Id, Client Secret, Access Token, and Refresh Token. 
```

Author: Done

> 4. First, we will create two defined sequences called `create.xml` and  `retrieve.xml` to create an Account and retrieve data. You can go to the source view of the XML configuration file of the API and copy the following configuration. Replace the `clientSecret`, `clientId`, `accessToken`, `refreshToken` with obtained values in step 3. 

Reviewer: ```suggestion
4. First, we will create two defined sequences called `create.xml` and  `retrieve.xml` to create an account and retrieve data. You can go to the source view of the XML configuration file of the API and copy the following configuration. Replace the `clientSecret`, `clientId`, `accessToken`, `refreshToken` with obtained values in step 3. 
```

Author: Done

> create.xml

Reviewer: Can we create this inline with the list number? The numbering breaks when we render this currently.

Author: Done

        <sObjectName>{$ctx:sObject}</sObjectName>
        <fieldAndValue>{$ctx:fieldAndValue}</fieldAndValue>
    </salesforcerest.create>
</sequence>
```

retrieve.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<sequence name="retrieve" trace="disable" xmlns="http://ws.apache.org/ns/synapse">
    <property expression="json-eval($.sObject)" name="sObject" scope="default" type="STRING"/>
    <property expression="json-eval($.fieldAndValue)" name="fieldAndValue" scope="default" type="STRING"/>
    <salesforcerest.init>
        <accessToken></accessToken>
        <apiVersion>v44.0</apiVersion>
        <hostName>https://login.salesforce.com</hostName>
        <refreshToken></refreshToken>
        <clientSecret></clientSecret>
        <clientId></clientId>
        <apiUrl>https://ap16.salesforce.com</apiUrl>
        <registryPath>connectors/SalesforceRest</registryPath>
    </salesforcerest.init>
    <salesforcerest.query>
        <queryString>select id, name from Account</queryString>
    </salesforcerest.query>
</sequence>
```

5. Now create the API **salesforcerest**. You can go to the source view of the XML configuration file of the API and copy the following configuration. 

> ```

Reviewer: Same comment as above.

Author: Done

<?xml version="1.0" encoding="UTF-8"?>
<api context="/salesforcerest" name="salesforcerest" xmlns="http://ws.apache.org/ns/synapse">
    <resource methods="POST">
        <inSequence>
            <sequence key="create"/>
            <sequence key="retrieve"/>
            <respond/>
        </inSequence>
        <outSequence/>
        <faultSequence/>
    </resource>
</api>
```

> 8. Follow these steps to export the artifacts. 

Reviewer: Update so that this is a different section. See File Connector Example.

Author: Done

{!references/connectors/exporting-artifacts.md !}

## Deployment
Follow these steps to deploy the exported CApp in the Enterprise Integrator Runtime. 
{!references/connectors/deploy-capp.md!}

## Testing
Save a file called data.json with the following payload. 
```
{
	"sObject":"Account",
	"fieldAndValue": {
    "name": "Engineers",
    "description":"This Account belongs to WSO2"
  }
}
```

> Invoke the API as shown below using the curl command. Curl Application can be downloaded from [here](https://curl.haxx.se/download.html).

Reviewer: ```suggestion
Invoke the API as shown below using the curl command. Curl application can be downloaded from [here](https://curl.haxx.se/download.html).
```

Author: Done

```
curl -X POST -d @data.json http://localhost:8280/salesforcerest --header "Content-Type:application/json"
```

> You will get set of Account names and their Ids as the output. 

Reviewer: ```suggestion
You will get a set of account names and the respective IDs as the output. 
```

Author: Done
